### PR TITLE
Search filters can be stored in Search and used to change filters on the back-end queries.

### PR DESCRIPTION
**Pull request by luk-kaminski**
Search filters can be stored in Search MongoDB document and used to change filters on the back-end queries.

## Description
This PR implements most back-end for for task #3387.
The main areas that are left to do are:
- integration tests,
- usage of filters in search/query content packs (if we believe it is needed).

From now on, it is possible to add and permanently store Search Filters in Queries inside each Search, under "filters" field.
Those Search Filters can be either "inlined" or "referenced".

Without Enterpise Plugin, this piece of functionality will have dummy implementations, that do nothing.
Many elements will be therefore implemented in a corresponding PR in graylog-plugin-enterprise.

## Motivation and Context
Front-end needs to store a relation between Searches and Search Filters.
Back-end needs to use Search Filters to actually filter the results accordingly.

## How Has This Been Tested?
As front-end is not yet implemented, it has been tested with REST API calls from "views/search" endpoint.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/views/ElasticsearchBackendTest.java [hunk 1]
```diff
@@ -27,41 +28,54 @@
...
+    private UsedSearchFiltersToQueryStringsMapper usedSearchFiltersToQueryStringsMapper;
 
-    @BeforeClass
-    public static void setup() {
+    @Before
+    public void setup() {
         Map<String, Provider<ESSearchTypeHandler<? extends SearchType>>> handlers = Maps.newHashMap();
         handlers.put(MessageList.NAME, () -> new ESMessageList(new QueryStringDecorators(Optional.empty())));
 
+        usedSearchFiltersToQueryStringsMapper = mock(UsedSearchFiltersToQueryStringsMapper.class);
```
Comment (todvora) [inline graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/views/ElasticsearchBackendTest.java:70]:
```suggestion
        usedSearchFiltersToQueryStringsMapper = usedSearchFilters -> Collections.emptySet();
```

File: graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendTest.java [hunk 2]
```diff
@@ -93,4 +110,48 @@ public void executesSearchForEmptySearchTypes() throws Exception {
...
+
+        final List<QueryBuilder> filters = ((BoolQueryBuilder) esQuery).filter();
+
+        //filter for empty ES query
+        assertThat(filters.stream()
+                .anyMatch(queryBuilder -> queryBuilder instanceof MatchAllQueryBuilder))
+                .isTrue();
+
+        //2 filters from search filters
+        assertThat(filters.stream()
```
Comment (todvora) [inline graylog-storage-elasticsearch7/src/test/java/org/graylog/storage/elasticsearch7/views/ElasticsearchBackendTest.java]: We have discussed some possible simplifications of these assertions. Could you please implement them?

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/module/SearchFiltersModule.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import org.graylog.plugins.views.search.searchfilters.db.SearchFiltersReFetcher;
+import org.graylog.plugins.views.search.searchfilters.db.UsedSearchFiltersToQueryStringsMapper;
+import org.graylog2.plugin.inject.Graylog2Module;
+
+public class SearchFiltersModule extends Graylog2Module {
+
+    @Override
+    protected void configure() {
+        super.configure();
+        OptionalBinder.newOptionalBinder(binder(), UsedSearchFiltersToQueryStringsMapper.class)
```
Comment (todvora) [inline graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/module/SearchFiltersModule.java:30]: Maybe reusing the newly created binder?

**Review by dennisoelkers — COMMENTED** [review]
Looks good so far, adding a couple of first questions.

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/UsedSearchFilter.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        property = ModelConstants.Fields.TYPE)
+@JsonSubTypes({
+        @JsonSubTypes.Type(value = InlineQueryStringSearchFilter.class, name = UsedSearchFilter.INLINE_QUERY_STRING_SEARCH_FILTER),
+        @JsonSubTypes.Type(value = ReferencedQueryStringSearchFilter.class, name = UsedSearchFilter.REFERENCED_SEARCH_FILTER),
+})
+public interface UsedSearchFilter {
```
[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/UsedSearchFilter.java:31]: It is also a slightly different POJO, the one supposed to be used on Search itself, not the one representing a document from a MongoDB collection.

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.plugins.views.search.searchfilters.model;
+
+public interface ModelConstants {
+
+    String MONGO_COLLECTION_NAME = "search_filters";
+
+    interface Fields {
+        String TYPE = "type";
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java:24]: How about moving them to `UsedSearchFilter` and use e.g. `UsedSearchfilter.title` from the subclasses? That is how we do it for other cases currently. If there is no benefit, I would keep it that way for consistency.

[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java:24]: Many constants are re-used in both SearchFilter and UseSearchFilter. If we want to move them away from this cental place, we would also have to duplicate them in both hierarchies.

[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java:24]: Afaics the only one which is used in both is `TYPE`. We tend to declare the type field for every entity separately anyhow. As mentioned before, I do not think there will be a need to change both at the same type (I cannot even recall a single case where we renamed the `type` field at all).

[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java:24]: All constants are re-used across both hierarchies, i.e. TITLE is used 12 times in the code, DESCRIPTION 10 times...
The question is - are you upset only with the TYPE field, or do you complain about the whole class with the constants and vote for its extermination? ;)

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/db/IgnoreSearchFilters.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Search filters are enterprise feature, so by default they won't be loaded.
+ *
+ * As long as proper enterprise module won't be loaded, this class will provide dummy/do-nothing implementations for Search Filter related functionality.
+ */
+public class IgnoreSearchFilters implements UsedSearchFiltersToQueryStringsMapper, SearchFiltersReFetcher {
```
[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/db/IgnoreSearchFilters.java:31]: I do not have a very strong opinion, but I am leaning towards this/existing approach. What I like about it is that we have "IgnoreSearchFilters" acting as a kind of "Null Object" and no client code has to implement any special handling of the optional dependency - we have already done it, once for all client code, current and possibly future. 
Feel free to point out any disadvantages of this approach.

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java:24]: I think we are already at the point of diminishing returns, as this is really not a big thing to me. What we have practised in the past was to centralize constants like this in the top-level interface/class of the entity and not share them between entities. I am always up for questioning the status quo when it's an improvement, but in this case I do not think we gain anything from it, but lose a bit of consistency.

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/db/IgnoreSearchFilters.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/db/IgnoreSearchFilters.java:31]: We've discussed it with Tomas, we are ok leaving it as it is.

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/module/SearchFiltersModule.java (hunk 3 above)
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/module/SearchFiltersModule.java:30]: We've figured out that those invocations are about two different types and cannot be reused.

File: graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/views/ElasticsearchBackendTest.java (hunk 1 above)
Comment (luk-kaminski) [reply graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/views/ElasticsearchBackendTest.java:70]: This change is problematic, as I'd have to re-init the whole "backend" class, initialized in setUp()... With mock I can just change the behavior in one test...

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/model/ModelConstants.java:24]: We have decided to exterminate the centralized constants. :)

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/db/UsedSearchFiltersToQueryStringsMapper.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.graylog.plugins.views.search.searchfilters.model.UsedSearchFilter;
+
+import java.util.Collection;
+import java.util.Set;
+
+@FunctionalInterface
+public interface UsedSearchFiltersToQueryStringsMapper {
+
+    Set<String> map(final Collection<UsedSearchFilter> usedSearchFilters);
```
[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/searchfilters/db/UsedSearchFiltersToQueryStringsMapper.java:27]: We decided not to refactor anything in this area for now...

File: graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/views/ElasticsearchBackendTest.java (hunk 1 above)
Comment (todvora) [reply graylog-storage-elasticsearch6/src/test/java/org/graylog/storage/elasticsearch6/views/ElasticsearchBackendTest.java:70]: Then don't worry about it and keep it as it is now :+1:

**Pull request merged**